File: LeetCode/542_01_Matrix.java
class Solution {
    public int[][] updateMatrix(int[][] matrix) {

        int numRow = matrix.length;
        int numCol = matrix[0].length;

        int[][] ans = new int[numRow][numCol];
        Queue<Pair> queue = new LinkedList<Pair>();

        for (int i = 0 ; i < numRow ; ++i) {
            for (int j = 0 ; j < numCol ; ++j) {
                if (matrix[i][j] == 0) {
                    ans[i][j] = 0;
                } else {
                    queue.add(new Pair(i, j));
                }
            }
        }

        while (!queue.isEmpty()) {
            Pair pair = queue.poll();
            int distance = calculateMinDistance(matrix, numRow, numCol, pair);
            ans[pair.row][pair.col] = distance;
        }

        return ans;
    }

    private int calculateMinDistance(int[][] matrix, int numRow, int numCol, Pair root) {

        int distance = 1;
        Queue<Tuple> queue = new LinkedList<Tuple>();
        queue.add(new Tuple(root.row, root.col, 1));

        while (!queue.isEmpty()) {
            Tuple tuple = queue.poll();
            int row = tuple.row;
            int col = tuple.col;
            distance = tuple.distance;

            if (row > 0) {
                if (matrix[row - 1][col] == 0) {
                    break;
                }
                queue.add(new Tuple(row - 1, col, distance + 1));
            }
            if (row < numRow - 1) {
                if (matrix[row + 1][col] == 0) {
                    break;
                }
                queue.add(new Tuple(row + 1, col, distance + 1));
            }
            if (col > 0) {
                if (matrix[row][col - 1] == 0) {
                    break;
                }
                queue.add(new Tuple(row, col - 1, distance + 1));
            }
            if (col < numCol - 1) {
                if (matrix[row][col + 1] == 0) {
                    break;
                }
                queue.add(new Tuple(row, col + 1, distance + 1));
            }
        }

        return distance;
    }

    private class Pair {
        public int row;
        public int col;

        public Pair(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }

    private class Tuple {
        public int row;
        public int col;
        public int distance;

        public Tuple(int row, int col, int distance) {
            this.row = row;
            this.col = col;
            this.distance = distance;
        }
    }
}

File: LeetCode/542_01_Matrix.cpp
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& matrix) {

        int num_row = matrix.size();
        int num_col = matrix[0].size();

        std::queue<std::pair<int, int>> queue;
        for (int i = 0 ; i < num_row ; ++i) {
            for (int j = 0 ; j < num_col ; ++j) {
                if (matrix[i][j] == 0) {
                    queue.push(std::make_pair(i, j));
                } else {
                    matrix[i][j] = INT_MAX;
                }
            }
        }

        std::vector<std::pair<int, int>> directions{
            {-1, 0},
            {1, 0},
            {0, -1},
            {0, 1}
        };

        while (!queue.empty()) {
            auto begin = queue.front();
            queue.pop();

            for (const auto& direction : directions) {
                int row = begin.first + direction.first;
                int col = begin.second + direction.second;

                if (row < 0 ||
                    row == num_row ||
                    col < 0 ||
                    col == num_col ||
                    matrix[row][col] <= matrix[begin.first][begin.second]) {
                    continue;
                }

                matrix[row][col] = matrix[begin.first][begin.second] + 1;
                queue.push(std::make_pair(row, col));
            }
        }

        return matrix;
    }
};
